### PR TITLE
Benchmark support

### PR DESCRIPTION
**Pull request by hidalgoss**
Please, review new changes.

### CODE REVIEW

File: examples/chapter1/CMakeLists.txt [hunk 1]
```diff
@@ -139,12 +162,15 @@ ADD_EXECUTABLE(${PROJ_UNIT_TESTS} ${${PROJ_NAME}_HEADERS} ${${PROJ_NAME}_SRC} )
...
-  gtest_main
+  benchmark
+  benchmark_main 
+  gtest
   )
 
 TARGET_INCLUDE_DIRECTORIES(${PROJ_NAME}_unit_tests
   PRIVATE
   ${CMAKE_CURRENT_SOURCE_DIR}/include 
+  /home/jose/klepsydra/klepsydra/development/community/install/include
```
Comment (pablogh-2000) [inline examples/chapter1/CMakeLists.txt:173]: no absolute path

File: examples/chapter1/CMakeLists.txt [hunk 2]
```diff
@@ -127,6 +127,29 @@ TARGET_LINK_LIBRARIES(${PROJ_NAME}_7
...
+    PUBLIC
+    ${KLEPSYDRA_CORE_LIBRARIES} 
+    benchmark
+    benchmark_main 
+    gtest
+    )
+TARGET_INCLUDE_DIRECTORIES(benchmark_example3
+  PUBLIC
+  ${CMAKE_CURRENT_SOURCE_DIR}/include 
+  /home/jose/klepsydra/klepsydra/development/community/install/include
```
Comment (pablogh-2000) [inline examples/chapter1/CMakeLists.txt:147]: no absolute local path

File: examples/chapter1/benchmark/benchmark_example3.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+//    1st range: poolsize of the smartpool.
+//    2nd range: vectorSize of the vector to be resized and published.
+//    3rd range: initialization bool. true-> initialize vector. false-> do not init vector.
+// ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ----
+static void BM_KCE_21_SmartPool_Test(benchmark::State &state)
+{
+    // Perform setup here
+    const int poolSize = state.range(0);
+    const int i_vectorSize(state.range(1));
+    const bool b_initialice = state.range(2);
```
Comment (pablogh-2000) [inline examples/chapter1/benchmark/benchmark_example3.cpp:40]: `b_initialize`

File: examples/chapter1/benchmark/benchmark_example3.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+    // This vector initialization seems be deterministic for lambda's resize() to be
+    // done in EventEmitterMiddlewareProvider() initialization. Must be created with
+    // the same size as the one to be resized? or Do we create it with no size, and
+    // lambda's resize SHOULD DO WHOLE WORK???
+    std::vector<float> vf_vec(i_vectorSize);
+    //std::vector<float> vf_vec;
+    std::function<void(std::vector<float> &)> fx_initializerFunction;
+    if (b_initialice) {
+        fx_initializerFunction = ([&](std::vector<float> &vf_vector) {
```
Comment (pablogh-2000) [inline examples/chapter1/benchmark/benchmark_example3.cpp:50]: replace `[&]` with `[i_vectorSize]`

File: examples/chapter1/benchmark/benchmark_example3.cpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+// 3) Ranges.
+//    1st range: poolsize of the smartpool.
+//    2nd range: vectorSize of the vector to be resized and published.
+//    3rd range: initialization bool. true-> initialize vector. false-> do not init vector.
+// ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ----
+static void BM_KCE_21_SmartPool_Test(benchmark::State &state)
+{
+    // Perform setup here
+    const int poolSize = state.range(0);
+    const int i_vectorSize(state.range(1));
```
Comment (pablogh-2000) [inline examples/chapter1/benchmark/benchmark_example3.cpp:39]: remove `i_` and `b_` we do not use this in our naming convention.

File: examples/chapter1/benchmark/benchmark_example3.cpp [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                                                                            fx_initializerFunction,
+                                                                            nullptr);
+    uint16_t ui_published_count = 0;
+    // This code gets timed
+    for (auto _ : state) {
+        vectorProvider.getPublisher()->publish(vf_vec);
+        ui_published_count++;
+    }
+    // end timed code.
+    state.counters["[poolSz]"] = poolSize;
```
Comment (pablogh-2000) [inline examples/chapter1/benchmark/benchmark_example3.cpp:68]: remove the internal brackets: 
`state.counters["poolSz"] = poolSize;`

File: examples/chapter1/benchmark/benchmark_example3.cpp [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    state.counters["[vectorSz]"] = i_vectorSize;
+    state.counters["[init]"] = b_initialice;
+    state.counters["[pblsh_cnt]"] = ui_published_count;
```
Comment (pablogh-2000) [inline examples/chapter1/benchmark/benchmark_example3.cpp:71]: use consistently camel case: `pblshCnt`

File: examples/chapter1/benchmark/benchmark_example3.cpp [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+BENCHMARK(BM_KCE_21_SmartPool_Test)
+    ->Name("hss BM_KCE_21_SmartPool_Test")
+    ->UseRealTime()
+    ->ArgsProduct({{0, 16, 32, 64}, {0, 512, 1024, 2048}, {true, false}});
```
Comment (pablogh-2000) [inline examples/chapter1/benchmark/benchmark_example3.cpp:77]: Add name to each range

File: examples/chapter1/include/sum_vector_data.h [hunk 9]
```diff
@@ -23,6 +23,7 @@
 #include <klepsydra/core/publisher.h>
 #include <klepsydra/core/subscriber.h>
 #include <numeric>
+#include <iostream>
 
 class SumVectorData
```
Comment (pablogh-2000) [inline examples/chapter1/include/sum_vector_data.h:28]: Remove this class. No needed

File: examples/chapter1/benchmark/benchmark_example3.cpp [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    kpsr::EventEmitterMiddlewareProvider<std::vector<float>> vectorProvider(nullptr,
+                                                                            "vector_provider",
+                                                                            poolSize,
+                                                                            fx_initializerFunction,
+                                                                            nullptr);
+    uint16_t ui_published_count = 0;
+    // This code gets timed
+    for (auto _ : state) {
+        vectorProvider.getPublisher()->publish(vf_vec);
+        ui_published_count++;
```
Comment (pablogh-2000) [inline examples/chapter1/benchmark/benchmark_example3.cpp:65]: nope. the counter increase should be done in the listener. This is very important

File: examples/chapter1/benchmark/benchmark_example3.cpp [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    } else {
+        fx_initializerFunction = nullptr;
+    }
+    kpsr::EventEmitterMiddlewareProvider<std::vector<float>> vectorProvider(nullptr,
+                                                                            "vector_provider",
+                                                                            poolSize,
+                                                                            fx_initializerFunction,
+                                                                            nullptr);
+    uint16_t ui_published_count = 0;
+    // This code gets timed
```
Comment (pablogh-2000) [inline examples/chapter1/benchmark/benchmark_example3.cpp:62]: Add listener before starting the loop. Look at the kpe-tutorial for examples if needed.

File: examples/chapter1/CMakeLists.txt (hunk 1 above)
Comment (hidalgoss) [reply examples/chapter1/CMakeLists.txt]: Done.

File: examples/chapter1/benchmark/benchmark_example3.cpp (hunk 7 above)
Comment (hidalgoss) [reply examples/chapter1/benchmark/benchmark_example3.cpp]: Done.

File: examples/chapter1/benchmark/benchmark_example3.cpp (hunk 8 above)
Comment (hidalgoss) [reply examples/chapter1/benchmark/benchmark_example3.cpp]: Still missing. User counters can be put in tabbed columns executing program with option:
/bin/benchmark_example3 **--benchmark_counters_tabular=true**
but still not found in google docs a trivial manner to rename column benchmarks ouputs.

File: examples/chapter1/benchmark/benchmark_example3.cpp (hunk 11 above)
Comment (hidalgoss) [reply examples/chapter1/benchmark/benchmark_example3.cpp:65]: Done.

File: examples/chapter1/include/sum_vector_data.h (hunk 9 above)
Comment (pablogh-2000) [reply examples/chapter1/include/sum_vector_data.h:28]: Why? I don't see it use anywhere.... I think we should remove it.

**Comment by pablogh-2000** [comment]
close this PR and delete branch.

**Pull request closed without merging**